File: backend/app/services/contribution_service.py
"""Contribution flow — debits user's main wallet, credits the pool, ledgers it.

All work happens inside the caller's session/transaction so balances cannot drift.
"""
from decimal import Decimal
from typing import Any

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ..enums import (
    ContributionStatus,
    NotificationType,
    TransactionDirection,
    TransactionType,
)
from ..errors import Errors
from ..models import Contribution, Notification, Pool, PoolMember, Transaction, User
from .pool_service import ensure_active_pool


async def make_contribution(session: AsyncSession, *, pool_id: str, user_id: str,
                            amount: str, description: str | None,
                            receipt_url: str | None) -> dict[str, Any]:
    pool = (await session.execute(select(Pool).where(Pool.id == pool_id))).scalar_one_or_none()
    if not pool:
        raise Errors.not_found("Pool")
    ensure_active_pool(pool)

    member = (await session.execute(
        select(PoolMember).where(PoolMember.poolId == pool_id, PoolMember.userId == user_id)
    )).scalar_one_or_none()
    if not member or not member.isActive:
        raise Errors.forbidden("Not a member of this pool")

    user = (await session.execute(select(User).where(User.id == user_id))).scalar_one_or_none()
    if not user:
        raise Errors.not_found("User")

    amt = Decimal(amount)
    if user.mainBalance < amt:
        raise Errors.insufficient_balance()

    user_before = user.mainBalance
    user_after = user_before - amt
    pool_before = pool.currentBalance
    pool_after = pool_before + amt

    user.mainBalance = user_after
    pool.currentBalance = pool_after

    contrib = Contribution(
        poolId=pool.id, userId=user.id, amount=amt,
        description=description, receiptUrl=receipt_url,
        status=ContributionStatus.COMPLETED,
    )
    session.add(contrib)
    await session.flush()

    session.add(Transaction(
        userId=user.id, type=TransactionType.CONTRIBUTION, direction=TransactionDirection.OUT,
        amount=amt, balanceBefore=user_before, balanceAfter=user_after,
        description=f"Contribution to {pool.name}",
        contributionId=contrib.id, metadata_={"poolId": pool.id},
    ))
    session.add(Transaction(
        poolId=pool.id, userId=user.id, type=TransactionType.CONTRIBUTION,
        direction=TransactionDirection.IN, amount=amt,
        balanceBefore=pool_before, balanceAfter=pool_after,
        description=f"Contribution from {user.displayName}", contributionId=contrib.id,
    ))

    others = (await session.execute(
        select(PoolMember.userId).where(
            PoolMember.poolId == pool.id, PoolMember.isActive.is_(True),
            PoolMember.userId != user.id,
        )
    )).scalars().all()
    for uid in others:
        session.add(Notification(
            userId=uid, type=NotificationType.CONTRIBUTION_RECEIVED,
            title="Contribution received",
            body=f"{user.displayName} contributed RM{amt:.2f} to {pool.name}.",
            metadata_={"poolId": pool.id, "contributionId": contrib.id, "amount": f"{amt:.2f}"},
        ))

    await session.commit()
    await session.refresh(contrib)
    return {"contribution": contrib, "poolBalance": pool_after, "userBalance": user_after}

File: src/services/contribution.service.ts
import { Prisma } from "@prisma/client";
import { prisma } from "../lib/prisma.js";
import { Errors } from "../utils/errors.js";
import { ensureActivePool } from "./pool.service.js";

/**
 * Atomically:
 *  1. Deducts amount from user's main wallet
 *  2. Adds amount to pool's currentBalance
 *  3. Records a Contribution row
 *  4. Records a Transaction (OUT for user)
 *  5. Records a Transaction (IN for pool, attributed to user)
 *  6. Notifies pool members
 *
 * All in a single Prisma transaction so balances cannot drift.
 */
export async function makeContribution(input: {
  poolId: string;
  userId: string;
  amount: string;
  description?: string;
  receiptUrl?: string;
}) {
  return prisma.$transaction(async (tx) => {
    const pool = await tx.pool.findUnique({ where: { id: input.poolId } });
    if (!pool) throw Errors.notFound("Pool");
    ensureActivePool(pool);

    const member = await tx.poolMember.findUnique({
      where: { poolId_userId: { poolId: input.poolId, userId: input.userId } },
    });
    if (!member || !member.isActive) throw Errors.forbidden("Not a member of this pool");

    const user = await tx.user.findUnique({ where: { id: input.userId } });
    if (!user) throw Errors.notFound("User");

    const amount = new Prisma.Decimal(input.amount);
    if (user.mainBalance.lessThan(amount)) throw Errors.insufficientBalance();

    const userBefore = user.mainBalance;
    const userAfter = userBefore.minus(amount);
    const poolBefore = pool.currentBalance;
    const poolAfter = poolBefore.plus(amount);

    await tx.user.update({ where: { id: user.id }, data: { mainBalance: userAfter } });
    await tx.pool.update({ where: { id: pool.id }, data: { currentBalance: poolAfter } });

    const contribution = await tx.contribution.create({
      data: {
        poolId: pool.id,
        userId: user.id,
        amount,
        description: input.description,
        receiptUrl: input.receiptUrl,
        status: "COMPLETED",
      },
    });

    await tx.transaction.create({
      data: {
        userId: user.id,
        type: "CONTRIBUTION",
        direction: "OUT",
        amount,
        balanceBefore: userBefore,
        balanceAfter: userAfter,
        description: `Contribution to ${pool.name}`,
        contributionId: contribution.id,
        metadata: { poolId: pool.id },
      },
    });

    await tx.transaction.create({
      data: {
        poolId: pool.id,
        userId: user.id,
        type: "CONTRIBUTION",
        direction: "IN",
        amount,
        balanceBefore: poolBefore,
        balanceAfter: poolAfter,
        description: `Contribution from ${user.displayName}`,
        contributionId: contribution.id,
      },
    });

    const otherMembers = await tx.poolMember.findMany({
      where: { poolId: pool.id, isActive: true, NOT: { userId: user.id } },
      select: { userId: true },
    });
    if (otherMembers.length > 0) {
      await tx.notification.createMany({
        data: otherMembers.map((m) => ({
          userId: m.userId,
          type: "CONTRIBUTION_RECEIVED" as const,
          title: "Contribution received",
          body: `${user.displayName} contributed RM${amount.toFixed(2)} to ${pool.name}.`,
          metadata: { poolId: pool.id, contributionId: contribution.id, amount: amount.toFixed(2) },
        })),
      });
    }

    return { contribution, poolBalance: poolAfter, userBalance: userAfter };
  });
}
